Accept a device ID to the login fallback endpoint.

clokep opened the pull request:
This modifies the login fallback page to accept a query parameter (`device_id`) which will get passed into the login endpoint during the login process.

This also reworks some of the `login.js` code to reduce duplicated code.

Fixes #5755

MSC: https://github.com/matrix-org/matrix-doc/pull/2604

## Questions

* Should we also accept an `initial_device_display_name` query parameter? It does not look any of the [other parameters](https://matrix.org/docs/spec/client_server/r0.6.1#post-matrix-client-r0-login) would be useful to accept since they're part of the login process.

clokep commented:
Also -- I wasn't really sure about the styling of the JavaScript code, but kind of just followed what was there.

clokep commented on synapse/static/client/login/js/login.js:
```diff
@@ -7,26 +7,19 @@ window.matrixLogin = {
-        token: loginToken
-    };
+    // Add the login type.
+    data.type = type;
```
This modifies the input value which is kind of meh, but I think this code is contained enough that it is OK + we control the values being sent to this function.

clokep commented on synapse/static/client/login/js/login.js:
```diff
@@ -50,8 +43,9 @@ var setFeedbackString = function(text) {
-    $("#sso_redirect_url").val(this_page);
+    // Set the redirect to come back to this page, a login token will get added
+    // and handled after the redirect.
+    $("#sso_redirect_url").val(window.location.href);
```
This change means that we use the whole URL as the redirect, which is necessary to pass the query parameters through SSO.

richvdh replied:
passing the query params through SSO feels like the sort of thing we shouldn't be doing, for fear of CSRF. can we stash them in a cookie or something instead?

clokep replied:
Hmm...the concern being the request will be modified before it gets back to us? I think that would be "bad", but I don't see a security issue with it. We can probably save it in a cookie though. I'll take a look!

clokep commented:
> Should we also accept an `initial_device_display_name` query parameter? It does not look any of the [other parameters](https://matrix.org/docs/spec/client_server/r0.6.1#post-matrix-client-r0-login) would be useful to accept since they're part of the login process.

Had a brief chat about this and it seems there's no reason not to include it. I've added it to this PR.

clokep commented:
@richvdh I modified this to save the query string parameters to a cookie in the case of SSO. (In the case of a password I don't think this is necessary. I'm hoping the code is clear, but for SSO the following happens:

* The query string is parsed to an object.
* The object is serialized to JSON.
* The JSON is stored into a cookie
* ✨ SSO occurs ✨ 
* The cookie is retrieved and de-JSON-ified

This let us keep the code paths similar to the password version where we just follow the first step (parse the query string to an object).

I also cleaned-up some more code and added a few more comments while messing around in here.

richvdh commented on synapse/static/client/login/js/login.js:
```diff
@@ -50,12 +61,19 @@ var setFeedbackString = function(text) {
+    // If inhibit_redirect is false, and SSO is the only supported login method,
+    // we can redirect straight to the SSO page.
     if (matrixLogin.serverAcceptsSso) {
+        // Before submitted SSO, set the current query parameters into a cookie
```
```suggestion
        // Before submitting SSO, set the current query parameters into a cookie
```

richvdh commented on synapse/static/client/login/js/login.js:
```diff
@@ -4,30 +4,41 @@ window.matrixLogin = {
-    }).fail(errorFunc);
-};
+// Titles get updated through the process to give users feedback.
+var TITLE_PRE_AUTH = "Log in with one of the following methods";
```
`const` is a thing in modern javascript https://caniuse.com/#feat=const

 :shrug:

clokep replied:
I debated about changing it... IE < 11 didn't support it and I didn't want to think too hard about our policy here. I'll do a separate PR that cleans up the style of this file a bit!
